File: lib/wordline.lua
-- The word line, read for the PDF: lib/wordline.py's grammar in Lua.
--
-- Loaded once by lib/frank-preamble.tex with dofile, so it is ordinary Lua:
-- none of the traps of code written inside \directlua (no # doubled, no %
-- taken for a comment, no backslash or brace eaten by TeX first) reach it,
-- and texlua runs it directly against tests/fixtures/wordline.json.
--
--   M.parse(line)      -> { {surface, reading}, ... }   or nil, code, message
--   M.align(fa, words) -> { {text, reading, k}, ... }   or nil, code, message
--   M.tex(fa, line)    -> the \FrankWord calls that set `fa` a word at a time
--   M.trailing(fa)     -> the punctuation that closes `fa`
--   M.aloud(reading, fa)     -> the chunk as the aloud pass reads it
--   M.tex_aloud(reading, fa) -> the same, with the reading in \FrankAloudText
--
-- k is a word's 0-based index, or false for whitespace between two words.
-- Codes are wordline.py's.  The text is drawn from `fa` and only the division
-- and the readings from the line, so a space the text holds is kept though the
-- line cannot spell it.  All three parsers compare code points without
-- normalising them, so a line that differs from its text only in
-- normalisation is refused by the checker before a build ever reaches here.

local M = {}

local SPACE = {}
for _, c in ipairs({9, 10, 11, 12, 13, 28, 29, 30, 31, 32, 0x85, 0xA0, 0x1680,
                    0x2028, 0x2029, 0x202F, 0x205F, 0x3000}) do
  SPACE[c] = true
end
for c = 0x2000, 0x200A do SPACE[c] = true end

local OPEN, CLOSE, FW_OPEN, FW_CLOSE = 40, 41, 0xFF08, 0xFF09

local function codepoints(s)
  local t = {}
  for _, c in utf8.codes(s) do t[#t + 1] = c end
  return t
end

local function text(t, i, j)
  local out = {}
  for k = i or 1, j or #t do out[#out + 1] = utf8.char(t[k]) end
  return table.concat(out)
end

local function fail(code, message)
  error({code = code, message = message}, 0)
end

local function collapse(t)
  local words, cur = {}, {}
  for _, c in ipairs(t) do
    if SPACE[c] then
      if #cur > 0 then words[#words + 1] = text(cur); cur = {} end
    else
      cur[#cur + 1] = c
    end
  end
  if #cur > 0 then words[#words + 1] = text(cur) end
  return table.concat(words, " ")
end

local function parse(line)
  local t = codepoints(line)
  local n = #t
  local words, surf, read = {}, {}, {}
  local state, start, i = 0, 1, 1        -- 0 surface, 1 reading, 2 closed

  local function shown(e)
    return "'" .. text(t, start, e - 1) .. "'"
  end

  local function finish(e)
    if state == 1 then fail("unclosed", shown(e) .. " opens a reading and never closes it") end
    if #surf > 0 then
      local r = collapse(read)
      if state == 2 and r == "" then
        fail("empty_reading", shown(e) .. " has empty parentheses: a word with no reading is written without them")
      end
      local wide = false
      for _, c in ipairs(surf) do if c == FW_OPEN then wide = true end end
      if wide and surf[#surf] == FW_CLOSE and surf[1] ~= FW_OPEN then
        fail("fullwidth", shown(e) .. " puts its reading in the fullwidth parentheses; a reading is written in ASCII ( )")
      end
      if surf[#surf] == OPEN or r:sub(1, 1) == "(" then
        fail("unwritable", shown(e) .. " ends in a literal (, which cannot be written back")
      end
      words[#words + 1] = {text(surf), r}
    end
    surf, read, state = {}, {}, 0
  end

  while i <= n do
    local c = t[i]
    if state == 2 then
      if not SPACE[c] then
        fail("joined", "a reading ends its word: a space is missing after " .. shown(i))
      end
      finish(i); i = i + 1; start = i
    elseif (c == OPEN or c == CLOSE) and t[i + 1] == c then
      if state == 1 then read[#read + 1] = c else surf[#surf + 1] = c end
      i = i + 2
    elseif state == 0 and SPACE[c] then
      finish(i); i = i + 1; start = i
    else
      if c == OPEN then
        if state == 1 then fail("second_open", "a second ( inside " .. shown(n + 1)) end
        if #surf == 0 then fail("no_word", "a reading with no word in front of it in " .. shown(n + 1)) end
        state = 1
      elseif c == CLOSE then
        if state == 0 then fail("stray_close", "a ) that closes nothing in " .. shown(n + 1)) end
        state = 2
      elseif state == 1 then
        read[#read + 1] = c
      else
        surf[#surf + 1] = c
      end
      i = i + 1
    end
  end
  finish(n + 1)
  return words
end

local function align(fa, words)
  local t = codepoints(fa)
  local n, i, out = #t, 1, {}
  local function mismatch()
    fail("reproduce", "the words do not reproduce the text '" .. fa .. "'")
  end
  for k, w in ipairs(words) do
    local j = i
    while j <= n and SPACE[t[j]] do j = j + 1 end
    if j > i then out[#out + 1] = {text(t, i, j - 1), "", false} end
    i = j
    local start = i
    for _, ch in utf8.codes(w[1]) do
      while i <= n and SPACE[t[i]] do i = i + 1 end
      if i > n or t[i] ~= ch then mismatch() end
      i = i + 1
    end
    out[#out + 1] = {text(t, start, i - 1), w[2], k - 1}
  end
  for k = i, n do
    if not SPACE[t[k]] then mismatch() end
  end
  if i <= n then out[#out + 1] = {text(t, i, n), "", false} end
  return out
end

local function guarded(f)
  return function(...)
    local ok, res = pcall(f, ...)
    if ok then return res end
    if type(res) == "table" then return nil, res.code, res.message end
    error(res, 0)
  end
end

M.parse = guarded(parse)
M.align = guarded(align)

-- The punctuation a chunk's text may close with: wordline.py's TRAIL, range
-- for range.  Declared before M.tex, which is the first to ask: a local
-- function is only in scope below its own line.
local TRAIL = {{0x21, 0x2F}, {0x3A, 0x40}, {0x5B, 0x60}, {0x7B, 0x7E}, {0xA1, 0xBF},
               {0x2010, 0x2027}, {0x2030, 0x205E}, {0x3001, 0x3003}, {0x3008, 0x3011},
               {0x3014, 0x301F}, {0x30FB, 0x30FB}, {0xFF01, 0xFF0F}, {0xFF1A, 0xFF20},
               {0xFF3B, 0xFF40}, {0xFF5B, 0xFF65}}

local function trails(c)
  if not c then return false end
  for _, r in ipairs(TRAIL) do
    if c >= r[1] and c <= r[2] then return true end
  end
  return false
end

-- What opens: a word ending in one of these is not parted from the next.
local OPENS = {}
for _, c in ipairs({0x28, 0x5B, 0x7B, 0xAB, 0x2018, 0x201C, 0x3008, 0x300A, 0x300C,
                    0x300E, 0x3010, 0x3014, 0x3016, 0x3018, 0x301A, 0x301D, 0xFF08,
                    0xFF3B, 0xFF5B, 0xFF5F, 0xFF62}) do
  OPENS[c] = true
end

-- The chunk's text as the calls that set it a word at a time: one
-- \FrankWord{text}{reading} per word, \FrankWordSep between them, and
-- whitespace from the text passed through as a space.  The backslash and the
-- braces are built from their codes, the way frank_breaks builds its
-- \allowbreak, so the answer is the same whatever reads this file.  A line
-- that does not fit the text is a TeX error that names the chunk: the build
-- stops rather than print a reading over the wrong characters.
-- No \FrankWordSep before a word that opens with closing punctuation (、 。 」),
-- nor after one that ends with an opening bracket: a word with its ruby is a
-- box, and the break the separator offers beside it is the one place a line
-- could begin with 。 -- it did, in the fixture edition, the moment 行きました
-- filled a line.  Without it babel keeps the two together as it keeps any two
-- characters its line-breaking rules forbid parting.
-- What may hang past the end of a line (burasagari): the comma and the full
-- stop, ideographic and fullwidth, and nothing else -- a bracket never hangs.
local HANG = {[0x3001] = true, [0x3002] = true, [0xFF0C] = true, [0xFF0E] = true}

function M.tex(fa, line)
  local bs, ob, cb = string.char(92), string.char(123), string.char(125)
  local words, code, message = M.parse(line)
  local spans
  if words then spans, code, message = M.align(fa, words) end
  if not spans then
    tex.error("the word line of the chunk '" .. fa .. "' cannot be set (" .. code .. "): " .. message)
    return fa
  end
  -- A bracket that opens and the marks that close ride with their word.  「 and
  -- 、 。 」 are words of their own in the line, but a line may neither open on
  -- 。 nor end on 「, so each is set inside the word beside it, as
  -- \FrankWordP{opening}{text}{reading}{closing}{hanging}, which measures them
  -- together (the preamble says why: a long word and its comma ran past the
  -- chunk column into the gloss).  {hanging} is a last 、 or 。 alone, the one mark the
  -- preamble may hang past the edge.  A word with neither is \FrankWord.
  local function every(t, test)
    if #t == 0 then return false end
    for _, c in ipairs(t) do if not test(c) then return false end end
    return true
  end
  local function opener(c) return OPENS[c] or false end
  local function closer(c) return trails(c) and not OPENS[c] end
  local items, pre = {}, nil
  local function flush_pre()
    if pre then items[#items + 1] = {pre = "", text = pre, reading = "", post = ""} end
    pre = nil
  end
  for _, s in ipairs(spans) do
    if s[3] == false then
      flush_pre()
      items[#items + 1] = " "
    else
      local t, bare = codepoints(s[1]), s[2] == ""
      if bare and every(t, closer) then
        flush_pre()
        local last = items[#items]
        if type(last) == "table" then
          last.post = last.post .. s[1]
        else
          items[#items + 1] = {pre = "", text = s[1], reading = "", post = ""}
        end
      elseif bare and every(t, opener) then
        pre = (pre or "") .. s[1]
      else
        items[#items + 1] = {pre = pre or "", text = s[1], reading = s[2], post = ""}
        pre = nil
      end
    end
  end
  flush_pre()
  local out, held = {}, false
  for _, it in ipairs(items) do
    if it == " " then
      out[#out + 1] = " "
    else
      local t = codepoints(it.pre .. it.text)
      if #out > 0 and out[#out] ~= " " and not held and not closer(t[1]) then
        out[#out + 1] = bs .. "FrankWordSep "
      end
      if it.pre == "" and it.post == "" then
        out[#out + 1] = bs .. "FrankWord" .. ob .. it.text .. cb .. ob .. it.reading .. cb
      else
        local p = codepoints(it.post)
        local k = #p
        if k > 0 and HANG[p[k]] then k = k - 1 end   -- one mark hangs, never two
        out[#out + 1] = bs .. "FrankWordP" .. ob .. it.pre .. cb .. ob .. it.text .. cb
                        .. ob .. it.reading .. cb .. ob .. text(p, 1, k) .. cb
                        .. ob .. text(p, k + 1) .. cb
      end
      local u = codepoints(it.text .. it.post)
      held = opener(u[#u])
    end
  end
  return table.concat(out)
end

-- The chunk read aloud: wordline.py's trailing() and aloud(), on the same
-- code points (TRAIL, above).  Whitespace inside the closing run is skipped,
-- and a reading that already ends with some of the run gets only the rest.
function M.trailing(fa)
  local t, out = codepoints(fa or ""), {}
  for i = #t, 1, -1 do
    if not SPACE[t[i]] then
      if not trails(t[i]) then break end
      table.insert(out, 1, t[i])
    end
  end
  return text(out)
end

-- (reading, the punctuation it lacks), or nil when there is no reading
local function aloud_parts(reading, fa)
  local r = codepoints(reading or "")
  local i, j = 1, #r
  while i <= j and SPACE[r[i]] do i = i + 1 end
  while j >= i and SPACE[r[j]] do j = j - 1 end
  if i > j then return nil end
  local p = codepoints(M.trailing(fa))
  local k = math.min(#p, j - i + 1)
  while k > 0 do
    local ends = true
    for m = 1, k do
      if r[j - k + m] ~= p[m] then ends = false; break end
    end
    if ends then break end
    k = k - 1
  end
  return text(r, i, j), text(p, k + 1, #p)
end

function M.aloud(reading, fa)
  local r, tail = aloud_parts(reading, fa)
  if not r then return fa or "" end
  return r .. tail
end

-- For \FrankAloud in the preamble: the reading goes into \FrankAloudText, the
-- language's wrapper (Chinese sets its pinyin in the roman), and the text's
-- punctuation stays outside it, in the language's own face.  That punctuation
-- is TeX source cut from the end of the text's, so it is written as what TeX
-- would PRINT: a character TeX reads as syntax becomes \char, and a closing }
-- or % cannot unbalance the pass; an escaped one, 50\% or \#, is the one
-- character it stands for, not a backslash and a percent; any other control
-- symbol (\, a thin space, \ a space) prints no character and is left out,
-- as ~, a space, is.  A chunk with no reading is its text, as \Strip prints it.
local SYNTAX = {[35] = true, [36] = true, [37] = true, [38] = true, [94] = true,
                [95] = true, [123] = true, [125] = true}           -- # $ % & ^ _ { }
local ESCAPED = {[35] = true, [36] = true, [37] = true, [38] = true,
                 [95] = true, [123] = true, [125] = true}          -- \# \$ \% \& \_ \{ \}

function M.tex_aloud(reading, fa)
  local bs, ob, cb = string.char(92), string.char(123), string.char(125)
  local r, tail = aloud_parts(reading, fa)
  if not r then return fa or "" end
  local out, t, i = {bs .. "FrankAloudText" .. ob .. r .. cb}, codepoints(tail), 1
  while i <= #t do
    local c, n = t[i], t[i + 1]
    if c == 92 then
      if n and ESCAPED[n] then out[#out + 1] = bs .. "char" .. n .. " " end
      -- \X with X ASCII is one control symbol; a backslash before anything
      -- else stood before the space the closing run skipped
      i = i + ((n and n < 128) and 2 or 1)
    else
      if SYNTAX[c] then
        out[#out + 1] = bs .. "char" .. c .. " "
      elseif c ~= 126 then
        out[#out + 1] = utf8.char(c)
      end
      i = i + 1
    end
  end
  return table.concat(out)
end

return M

File: lib/wordline.js
// SPDX-License-Identifier: GPL-3.0-or-later
/* The word line, read in the page: lib/wordline.py's grammar in JavaScript.
 *
 *   山(やま) へ 柴刈り(しばかり) に 、
 *
 * Both readers load this beside parseh.js.  The book reader and the player
 * draw a chunk a word at a time from it, the word strip edits it, and check()
 * runs under the strip as the person types, so the server's refusal is never
 * the first they hear of a mistake.  lib/wordline.py is the authority and
 * lib/wordline.lua the PDF's copy; tests/fixtures/wordline.json holds all
 * three to the same codes.
 *
 *   ParsehWordline.parse(line)        -> [[surface, reading], ...]  (throws {code, message})
 *   ParsehWordline.render(words)      -> line
 *   ParsehWordline.align(fa, words)   -> [[text, reading, k|null], ...]
 *   ParsehWordline.check(fa, line, lang, reading, reorders, door)
 *                                     -> {errors: [{code, message}], warnings: [...]}
 *   ParsehWordline.key(surface, reading) -> the word's identity, its own token
 *   ParsehWordline.aloud(reading, fa) -> the chunk as the aloud pass reads it
 *
 * `lang` is the registry record a page embeds (LANG, CFG.lang).  Everything
 * iterates by code point, never by UTF-16 unit, so a character outside the
 * basic plane is one character here as it is in Python.
 */
(function (root) {
  'use strict';

  // Python's str.isspace, exactly: JavaScript's \s is not the same set.
  var SPACE = {};
  [9, 10, 11, 12, 13, 28, 29, 30, 31, 32, 0x85, 0xA0, 0x1680,
   0x2028, 0x2029, 0x202F, 0x205F, 0x3000].forEach(function (c) { SPACE[c] = true; });
  for (var s0 = 0x2000; s0 <= 0x200A; s0++) SPACE[s0] = true;
  function isSpace(ch) { return !!SPACE[ch.codePointAt(0)]; }

  var TEX_SPECIALS = '\\{}$%&#_^~';
  var NOT_TEXT = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]/;
  var HAN = /[々〆㐀-䶿一-鿿豈-﫿\u{20000}-\u{3FFFF}]/u;
  var SCRIPT = /[ぁ-ゖゝゞァ-ヺー々〆㐀-䶿一-鿿豈-﫿\u{20000}-\u{3FFFF}]/u;
  var HIRA_TAIL = /[ぁ-ゖゝゞ]+$/;
  var HIRA_RUN = /^[ぁ-ゖゝゞ]+$/;
  var PUNCT = /\p{P}/u;

  function WordsError(code, message) {
    var e = new Error(message);
    e.code = code;
    return e;
  }

  function words(s) {                     // str.split()
    var out = [], cur = '';
    Array.from(s || '').forEach(function (ch) {
      if (isSpace(ch)) { if (cur) { out.push(cur); cur = ''; } }
      else cur += ch;
    });
    if (cur) out.push(cur);
    return out;
  }

  function parse(line) {
    if (typeof line !== 'string') throw WordsError('type', 'a word line is text');
    var t = Array.from(line), n = t.length;
    var pairs = [], surf = '', read = '', state = 0, start = 0, i = 0;

    function shown(end) { return "'" + t.slice(start, end).join('').trim() + "'"; }
    function finish(end) {
      if (state === 1) throw WordsError('unclosed', shown(end) + ' opens a reading and never closes it');
      var r = words(read).join(' ');
      if (surf) {
        if (state === 2 && !r)
          throw WordsError('empty_reading', shown(end) + ' has empty parentheses: a word with no reading is written without them');
        if (surf.indexOf('（') >= 0 && surf.slice(-1) === '）' && surf[0] !== '（')
          throw WordsError('fullwidth', shown(end) + ' puts its reading in the fullwidth parentheses （ ）: a reading is written in the ASCII ( ), and the fullwidth pair is part of the text');
        if (surf.slice(-1) === '(' || r[0] === '(')
          throw WordsError('unwritable', shown(end) + " ends in a literal '(', which cannot be written back: a word may not end in one nor a reading begin with one");
        pairs.push([surf, r]);
      }
      surf = ''; read = ''; state = 0;
    }

    while (i < n) {
      var c = t[i];
      if (state === 2) {
        if (isSpace(c)) { finish(i); i++; start = i; continue; }
        var j = i;
        while (j < n && !isSpace(t[j])) j++;
        var tail = t.slice(i, j).join('');
        if (HIRA_RUN.test(tail))
          throw WordsError('joined', shown(i) + ' is followed by ' + tail + ' with no space: if ' + tail +
            ' belongs to the word the reading covers it too, ' + surf + tail + '(' + read + tail +
            '), and if it is a word of its own put a space before it');
        throw WordsError('joined', 'a reading ends its word: put a space between ' + shown(i) + " and '" + tail + "'");
      }
      if ((c === '(' || c === ')') && t[i + 1] === c) {
        if (state === 1) read += c; else surf += c;
        i += 2;
        continue;
      }
      if (state === 0 && isSpace(c)) { finish(i); i++; start = i; continue; }
      if (c === '(') {
        if (state === 1) throw WordsError('second_open', 'a second ( inside ' + shown(n) + ': a literal parenthesis is written twice, ((');
        if (!surf) throw WordsError('no_word', 'a reading with no word in front of it, in ' + shown(n));
        state = 1;
      } else if (c === ')') {
        if (state === 0) throw WordsError('stray_close', shown(n) + ' closes a parenthesis it never opened: a literal one is written twice, ))');
        state = 2;
      } else if (state === 1) read += c;
      else surf += c;
      i++;
    }
    finish(n);
    return pairs;
  }

  function esc(s) { return s.replace(/\(/g, '((').replace(/\)/g, '))'); }

  function token(surface, reading) {
    reading = words(reading || '').join(' ');
    if (!surface) throw WordsError('no_text', 'a word with no text');
    if (Array.from(surface).some(isSpace))
      throw WordsError('space_in_word', "'" + surface + "' has a space in it: words are parted by spaces, so no word can hold one");
    if (surface.slice(-1) === '(' || reading[0] === '(')
      throw WordsError('unwritable', "'" + surface + "' cannot be written: a word may not end in '(' nor a reading begin with one");
    return esc(surface) + (reading ? '(' + esc(reading) + ')' : '');
  }

  function render(pairs) {
    return pairs.map(function (p) { return token(p[0], p[1]); }).join(' ');
  }

  function align(fa, pairs) {
    var t = Array.from(fa || ''), n = t.length, i = 0, out = [];
    function mismatch() {
      throw WordsError('reproduce', 'the words do not reproduce the text: the words give ' +
        pairs.map(function (p) { return p[0]; }).join('') + ' and the text is ' + t.join(''));
    }
    pairs.forEach(function (p, k) {
      var j = i;
      while (j < n && isSpace(t[j])) j++;
      if (j > i) out.push([t.slice(i, j).join(''), '', null]);
      i = j;
      var begin = i;
      Array.from(p[0]).forEach(function (ch) {
        while (i < n && isSpace(t[i])) i++;
        if (i >= n || t[i] !== ch) mismatch();
        i++;
      });
      out.push([t.slice(begin, i).join(''), p[1], k]);
    });
    for (var k = i; k < n; k++) if (!isSpace(t[k])) mismatch();
    if (i < n) out.push([t.slice(i).join(''), '', null]);
    return out;
  }

  // The chunk read aloud: wordline.py's trailing() and aloud(), on the same
  // code points (TRAIL), whitespace inside the closing run skipped.
  var TRAIL = [[0x21, 0x2F], [0x3A, 0x40], [0x5B, 0x60], [0x7B, 0x7E], [0xA1, 0xBF],
               [0x2010, 0x2027], [0x2030, 0x205E], [0x3001, 0x3003], [0x3008, 0x3011],
               [0x3014, 0x301F], [0x30FB, 0x30FB], [0xFF01, 0xFF0F], [0xFF1A, 0xFF20],
               [0xFF3B, 0xFF40], [0xFF5B, 0xFF65]];
  function trails(ch) {
    var c = ch.codePointAt(0);
    return TRAIL.some(function (r) { return c >= r[0] && c <= r[1]; });
  }

  function trailing(fa) {
    var t = Array.from(fa || ''), out = [];
    for (var i = t.length - 1; i >= 0; i--) {
      if (isSpace(t[i])) continue;
      if (!trails(t[i])) break;
      out.unshift(t[i]);
    }
    return out.join('');
  }

  function aloud(reading, fa) {
    var r = Array.from(reading || ''), i = 0, j = r.length;
    while (i < j && isSpace(r[i])) i++;
    while (j > i && isSpace(r[j - 1])) j--;
    if (i >= j) return fa || '';
    var s = r.slice(i, j).join(''), p = Array.from(trailing(fa)), k = p.length;
    while (k && !s.endsWith(p.slice(0, k).join(''))) k--;
    return s + p.slice(k).join('');
  }

  function sounds(pairs) {
    return pairs.filter(function (p) { return p[1] || !HAN.test(p[0]); })
                .map(function (p) { return p[1] || p[0]; });
  }

  function kanaKey(s) {
    return Array.from(s.normalize('NFC')).filter(function (ch) {
      return !isSpace(ch) && !PUNCT.test(ch);
    }).join('');
  }

  function romanKey(s) {
    return s.normalize('NFD').toLowerCase().replace(/[^a-z0-9]/g, '');
  }

  function agrees(pairs, reading, lang) {
    if (!(reading || '').trim() || !pairs.length) return null;
    var key = lang.reading ? kanaKey : romanKey;
    return key(sounds(pairs).join('')) === key(reading);
  }

  // code point by code point, never normalised: wordline.py's norm() says why
  function norm(s, lang) {
    s = s || '';
    if (lang.strip) s = s.replace(new RegExp('[' + lang.strip + ']', 'g'), '');
    return lang.word_sep ? words(s).join(' ') : words(s).join('');
  }

  function diff(got, want) {
    var a = Array.from(got), b = Array.from(want), k = 0;
    while (k < a.length && k < b.length && a[k] === b[k]) k++;
    function at(x) { return k < x.length ? 'U+' + x[k].codePointAt(0).toString(16).toUpperCase().padStart(4, '0') + ' ' + x[k] : 'nothing'; }
    return 'the words give ' + got + ' and the text is ' + want + ': at character ' + k +
      ' the words have ' + at(a) + ' and the text has ' + at(b);
  }

  function check(fa, line, lang, reading, reorders, door) {
    var errors = [], warnings = [];
    function note(list, code, message) { list.push({code: code, message: message}); }
    door = door || 'book';
    if (typeof line !== 'string') { note(errors, 'type', 'words must be text'); return {errors: errors, warnings: warnings}; }
    if (!lang.words) { note(errors, 'no_layer', lang.name + ' has no word layer, so a chunk carries no words'); return {errors: errors, warnings: warnings}; }
    var m = NOT_TEXT.exec(line);
    if (m) note(errors, 'control', 'words carries a character that is not text, at character ' + m.index);
    if (door === 'book') {
      var bad = Array.from(new Set(Array.from(line).filter(function (ch) { return TEX_SPECIALS.indexOf(ch) >= 0; })));
      if (bad.length) note(errors, 'tex_special', 'words may not contain ' + bad.join(' ') + ': LaTeX reads it as an instruction');
      if (/\n[ \t]*\n/.test(line)) note(errors, 'blank_line', 'a blank line inside words ends the paragraph');
    }
    if (errors.length) return {errors: errors, warnings: warnings};
    var pairs;
    try { pairs = parse(line); }
    catch (e) { note(errors, e.code || 'parse', e.message); return {errors: errors, warnings: warnings}; }
    if (!pairs.length) { note(errors, 'empty', 'words is empty: a chunk without words is written without the field'); return {errors: errors, warnings: warnings}; }
    var got = norm(pairs.map(function (p) { return p[0]; }).join(lang.word_sep || ''), lang), want = norm(fa, lang);
    if (got !== want) { note(errors, 'reproduce', 'the words do not reproduce the text: ' + diff(got, want)); return {errors: errors, warnings: warnings}; }

    var isReading = lang.reading_chars ? new RegExp('^[' + lang.reading_chars + '\\s]+$') : null;
    var hasScript = lang.chars ? new RegExp('[' + lang.chars + ']') : null;
    pairs.forEach(function (p) {
      var s = p[0], r = p[1];
      if (!r && HAN.test(s) && !reorders) note(warnings, 'no_reading', s + ' has no reading');
      if (r && lang.reading && !(isReading && isReading.test(r.trim())))
        note(warnings, 'reading_script', 'the reading of ' + s + ', ' + r + ', is not written in ' + (lang.reading_label || "the reading's script"));
      if (r && !lang.reading && hasScript && hasScript.test(r))
        note(warnings, 'reading_script', 'the reading of ' + s + ', ' + r + ', has ' + lang.name + ' characters in it');
      var tail = HIRA_TAIL.exec(s);
      if (lang.reading && r && tail && HAN.test(s) && r.slice(-tail[0].length) !== tail[0])
        note(warnings, 'okurigana', s + '(' + r + '): the reading should cover the whole word, ending with its okurigana ' + tail[0]);
      if (SCRIPT.test(s) && Array.from(s).some(function (ch) { return PUNCT.test(ch); }))
        note(warnings, 'punctuation', s + ' runs punctuation into a word: the punctuation is a word of its own');
    });
    if (!reorders && agrees(pairs, reading, lang) === false)
      note(warnings, 'disagrees', 'the words read ' + sounds(pairs).join(lang.reading ? '' : ' ') +
        " and the chunk's reading is " + reading + ': a text read out of its written order (kanbun) is marked "reorders": true, and anything else is a reading to correct');
    return {errors: errors, warnings: warnings};
  }

  root.ParsehWordline = {
    parse: parse, render: render, token: token, key: token, align: align,
    check: check, sounds: sounds, agrees: agrees, HAN: HAN,
    trailing: trailing, aloud: aloud
  };
})(typeof globalThis !== 'undefined' ? globalThis : this);
